modules/log.cpp

```diff
@@ -551,6 +555,74 @@ CModule::EModRet CLogMod::OnChanMsg(CNick& Nick, CChan& Channel,
     return CONTINUE;
 }
 
+bool CLogMod::OnWebRequest(CWebSock& WebSock, const CString& sPageName, CTemplate& Tmpl) {
+    if(GetType() == CModInfo::EModuleType::GlobalModule) {
+        return true;
+    }
+
+    CString sPath(WebSock.GetParam("path", false));
+    CString sFullPath(CDir::CheckPathPrefix(GetSavePath(), sPath));
+
+    if (sFullPath.empty()) {
+        sPath = "";
+        sFullPath = GetSavePath();
+    }
+
+    CFile File(sFullPath);
+    CString sDir(File.IsDir() ? sFullPath : File.GetDir());
+    CDir Dir(sDir);
+    CString sPrefix(sDir);
+
+    sPrefix.TrimPrefix(GetSavePath());
+    sPrefix.TrimPrefix("/");
+    if (!sPrefix.empty() && !sPrefix.EndsWith("/"))
+        sPrefix += "/";
+
+    if (!sPath.empty()) {
+        CTemplate& Row = Tmpl.AddRow("Files");
+        Row["Short"] = "..";
+        Row["Long"] = sPrefix + "..";
+        Row["Dir"] = "true";
+    }
+
+    for (const CFile* pFile : Dir) {
+        CTemplate& Row = Tmpl.AddRow("Files");
+        Row["Short"] = pFile->GetShortName();
+        Row["Long"] = sPrefix + pFile->GetShortName();
+        if (pFile->IsDir()) {
+            Row["Dir"] = "true";
+        } else {
+            Row["File"] = "true";
+        }
+    }
+
+    const size_t LINES_PER_PAGE = 512;
+    size_t Page = 0;
+    WebSock.GetParam("page", false).Convert(&Page);
+    const size_t Lines = Page * LINES_PER_PAGE;
+
+    if (File.IsReg()) {
+        CString Line;
+        File.Open();
+        for (size_t i = 0; i < Lines && File.ReadLine(Line); i++)
```

Comment:
If user has to go through pages, the file will be read again and again. Maybe it's easier to show the whole file without paging?

Reply:
That was what I originally wanted to do but the comments on the first attempt to implement this raised concerns about large log files (> 512KB) so I thought it would be better to page it by line numbers. If you would prefer it that way I can change it back and remove the paging parts from the template.

Reply:
I remember it now, yes.

Probably showing logs in web interface should have a better source than a plain text file, but something where you can say LIMIT and OFFSET like in SQL…

Reply:
But the plain text file is what is actually written. SQL isn't an option for this without changing how logs are managed. And since the files are line oriented I have to perform manual seeking since the lines can end pretty much anywhere, rendering byte offsets worthless except for forward paging only. Though I could make the lines per page configurable as a GET parameter if this would help. The way I had it setup was the best way I could think of to keep the memory usage from getting out of hand.

What would you have me do?

Reply:
What about counting bytes instead of lines per page? And when it got in a middle of a line, round to the nearest line ending, which should be near enough. Then byte offset could be set via GET

Reply:
That would work more or less but it has problems for "going back" as there's no way to know where the previous page positions were. But it works fine for going forward in one direction. I think we could leave "previous" as a function of the browser back button perhaps?

Reply:
Offsets of previous pages could be passed through browser in a hidden field?

Reply:
That would require a switch to POST requests since GET can't really handle that much data.

Reply:
Ok. This has been implemented as the new pagination method for the log files.